components/org.wso2.carbon.uis/src/main/java/org/wso2/carbon/uis/internal/io/deployment/ArtifactAppDeployer.java

```diff
@@ -27,125 +27,170 @@
         LOGGER.debug("An instance of class '{}' registered as an app deployment listener.",
                      listener.getClass().getName());
     }
 
     protected void unregisterListener(AppDeploymentEventListener listener) {
-        this.eventListeners.remove(listener);
+        this.appDeploymentEventListener = null;
         LOGGER.debug("An instance of class '{}' unregistered as an app deployment listener.",
                      listener.getClass().getName());
     }
 
     @Activate
     protected void activate(BundleContext bundleContext) {
-        start();
         LOGGER.debug("Carbon UI server app deployer activated.");
     }
 
     @Deactivate
     protected void deactivate(BundleContext bundleContext) {
-        stop();
+        try {
+            for (String artifactKey : deployedApps.keySet()) {
+                undeploy(artifactKey);
+            }
+        } catch (CarbonDeploymentException e) {
+            throw new UISRuntimeException("An error occurred during deactivation of app deployer.", e);
+        }
         LOGGER.debug("Carbon UI server app deployer deactivated.");
     }
 
-    private void start() {
-        if (!Files.exists(appsRepository)) {
-            LOGGER.debug("Web apps repository '{}' does not exists.", appsRepository.toString());
-            return;
-        }
+    @Override
+    public void init() {
+        LOGGER.debug("Carbon UI server app deployer initialized.");
+    }
 
-        Set<AppReference> appReferences;
-        try {
-            appReferences = Files.list(appsRepository)
-                    .filter(Files::isDirectory)
-                    .map(ArtifactAppReference::new)
-                    .collect(Collectors.toSet());
-        } catch (IOException e) {
-            throw new DeploymentException("Cannot list web apps in '" + appsRepository + "' directory.", e);
+    @Override
+    public Object deploy(Artifact artifact) throws CarbonDeploymentException {
+        Path appPath = artifact.getFile().toPath();
+        if (!isValidAppArtifact(appPath)) {
+            throw new DeploymentException("Artifact located in '" + appPath + "'is not a valid web app.");
         }
-        publishAppsDeploymentEvents(appReferences);
+
+        App app = createApp(appPath);
+        String artifactKey = "webapp:" + app.getName();
+
+        appDeploymentEventListener.appDeploymentEvent(app);
+        deployedApps.put(artifactKey, app);
+        LOGGER.debug("Web app '{}' deployed for context path '{}'.", app.getName(), app.getContextPath());
+        return artifactKey;
     }
 
-    private void stop() {
-        for (String appName : deployedAppNames) {
-            publishAppUndeploymentEvent(appName);
+    @Override
+    public void undeploy(Object key) throws CarbonDeploymentException {
+        App app = deployedApps.remove(key.toString());
+        if (app == null) {
+            throw new CarbonDeploymentException("Web app for key '{}' cannot be found to undeploy.");
```

Comment:
{} ?

Reply:
SLF4J supports parameterized log messages. See https://www.slf4j.org/faq.html#logging_performance

Reply:
Yes, but you haven't specified the parameter's value right?

Reply:
Sorry, this is not a log message, rather its a exception message.
Fixed in commit b82e5a5642019c8bbeb38aac13d6e471132d653a